### PR TITLE
Transaction Storage Engine Update

### PR DESCRIPTION
**Pull request by ejfitzgerald**

### CODE REVIEW

File: libs/ledger/include/ledger/storage_unit/recent_transaction_cache.hpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * The recent transactions cache is a fixed size of transactions that can be collected periodically
+ * by the miner to be incorporated into the subsequent blocks
+ */
+class RecentTransactionsCache
+{
+public:
+  using TxLayouts = std::vector<chain::TransactionLayout>;
+
+  explicit RecentTransactionsCache(std::size_t max_cache_size, uint32_t log2_num_lanes);
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/recent_transaction_cache.hpp]: unneccessary explicit here

File: libs/ledger/include/ledger/storage_unit/recent_transaction_cache.hpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+namespace chain {
+
+class Transaction;
+
+}  // namespace chain
+
+namespace ledger {
+
+/**
+ * The recent transactions cache is a fixed size of transactions that can be collected periodically
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/recent_transaction_cache.hpp:38]: 👍

File: libs/ledger/include/ledger/storage_unit/transaction_archiver.hpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+#include "core/state_machine.hpp"
+
+namespace fetch {
+namespace ledger {
+
+class TransactionPoolInterface;
+class TransactionStoreInterface;
+
+/**
+ * The transaction archiver manages transactions between a pool and a store. Once a transaction
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/transaction_archiver.hpp:31]: 👍

File: libs/ledger/include/ledger/storage_unit/transaction_finder_protocol.hpp [hunk 4]
```diff
@@ -55,19 +56,19 @@ class TxFinderProtocol : public fetch::service::Protocol
    * @param rid the Resource ID output
    * @return true if a Resource ID had been retrieved from the queue, false if the queue was empty
    */
-  bool Pop(storage::ResourceID &rid);
+  bool Pop(Digest &digest);
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/transaction_finder_protocol.hpp:59]: was there some bug here?

File: libs/ledger/include/ledger/storage_unit/transaction_pool_interface.hpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+namespace fetch {
+namespace ledger {
+
+class TransactionPoolInterface : public TransactionStoreInterface
+{
+public:
+  TransactionPoolInterface()           = default;
+  ~TransactionPoolInterface() override = default;
+
+  /// @name Transaction Pool Interface
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/transaction_pool_interface.hpp:31]: delete?

File: libs/ledger/include/ledger/storage_unit/transaction_storage_engine.hpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+namespace fetch {
+namespace core {
+
+class Reactor;
+
+}  // namespace core
+
+namespace ledger {
+
+class TransactionStorageEngine : public TransactionStorageEngineInterface
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/transaction_storage_engine.hpp:39]: Start your engines!

File: libs/ledger/include/ledger/storage_unit/transaction_storage_engine_interface.hpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+{
+public:
+  using TxArray   = std::vector<chain::Transaction>;
+  using TxLayouts = std::vector<chain::TransactionLayout>;
+
+  // Construction / Destruction
+  TransactionStorageEngineInterface()          = default;
+  virtual ~TransactionStorageEngineInterface() = default;
+
+  /// @name Transaction Storage Interface
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/transaction_storage_engine_interface.hpp:37]: delete?

File: libs/ledger/include/ledger/storage_unit/transaction_store.hpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  /// @{
+  void     Add(chain::Transaction const &tx) override;
+  bool     Has(Digest const &tx_digest) const override;
+  bool     Get(Digest const &tx_digest, chain::Transaction &tx) const override;
+  uint64_t GetCount() const override;
+  /// @}
+
+  /// @mame Low Level Subtree Access
+  /// @{
+  TxArray PullSubtree(Digest const &partial_digest, uint64_t bit_count, uint64_t pull_limit);
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/transaction_store.hpp:55]: Do we  know the cause of the subtree bug?

File: libs/ledger/include/ledger/storage_unit/transaction_store_aggregator.hpp [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "ledger/storage_unit/transaction_store_interface.hpp"
+
+namespace fetch {
+namespace ledger {
+
+class TransactionStoreAggregator : public TransactionStoreInterface
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/transaction_store_aggregator.hpp:33]: class level comment? What does this do

File: libs/ledger/src/storage_unit/lane_service.cpp [hunk 10]
```diff
@@ -170,7 +171,7 @@ void LaneService::Stop()
   state_db_protocol_.reset();
   state_db_.reset();
   tx_store_protocol_.reset();
-  tx_store_.reset();
+  //  tx_store_.reset();
```
Comment (n-hutton) [inline libs/ledger/src/storage_unit/lane_service.cpp]: delete?

File: libs/ledger/src/storage_unit/transaction_storage_protocol.cpp [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+  Expose(GET_RECENT, this, &TransactionStorageProtocol::GetRecent);
+}
+
+/**
+ * Create a total metric for a specified operation
+ *
+ * @param operation The operation
+ * @return The generated counter
+ */
+telemetry::CounterPtr TransactionStorageProtocol::CreateCounter(char const *operation) const
```
Comment (n-hutton) [inline libs/ledger/src/storage_unit/transaction_storage_protocol.cpp:75]: Why create a function for this? Could it be common code somewhere?

File: libs/ledger/src/storage_unit/transaction_store.cpp [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+TxArray TransactionStore::PullSubtree(Digest const &partial_digest, uint64_t bit_count,
+                                      uint64_t pull_limit)
+{
+  TxArray ret{};
+
+  uint64_t counter = 0;
+
+  archive_.Flush(false);
+  archive_.WithLock([this, &pull_limit, &ret, &counter, &partial_digest, bit_count]() {
+    // This is effectively saying get all objects whose ID begins rid & mask
```
Comment (n-hutton) [inline libs/ledger/src/storage_unit/transaction_store.cpp:136]: 👍

File: libs/ledger/src/storage_unit/transaction_store_sync_protocol.cpp [hunk 13]
```diff
@@ -151,18 +167,43 @@ TransactionStoreSyncProtocol::TxArray TransactionStoreSyncProtocol::PullObjects(
   return ret;
 }
 
-TransactionStoreSyncProtocol::TxArray TransactionStoreSyncProtocol::PullSpecificObjects(
-    std::vector<storage::ResourceID> const &rids)
+/**
```
Comment (n-hutton) [inline libs/ledger/src/storage_unit/transaction_store_sync_protocol.cpp:170]: Can you summarise what major changes there are to TX sync?

File: libs/ledger/tests/unit/storage/transaction_archiver_tests.cpp [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    FAIL() << "State machine never completed";
+  }
+
+  NiceMock<MockTransactionPool>  pool_;
+  NiceMock<MockTransactionStore> store_;
+  TransactionGenerator           tx_gen_;
+  TransactionArchiver            archiver_{pool_, store_};
+};
+
+MATCHER_P(IsTransaction, digest, "")  // NOLINT
```
Comment (n-hutton) [inline libs/ledger/tests/unit/storage/transaction_archiver_tests.cpp:61]: What happens if you don't nolint this?

File: libs/ledger/tests/unit/storage/transaction_generator.hpp [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+        .TargetChainCode("foo.bar.baz", BitVector{})
+        .Action("test")
+        .Data(GenerateRandomData())
+        .Signer(public_key_)
+        .Seal()
+        .Sign(private_key_)
+        .Build();
+  }
+
+  Txs GenerateRandomTxs(std::size_t count)
```
Comment (n-hutton) [inline libs/ledger/tests/unit/storage/transaction_generator.hpp:51]: I recall there are other fns floating around that generate txs - can you delete them?

File: libs/muddle/include/muddle/punishment_broadcast_channel.hpp [hunk 16]
```diff
@@ -115,6 +115,8 @@ class PunishmentBroadcastChannel : public service::Protocol, public BroadcastCha
   std::weak_ptr<core::Runnable> GetRunnable() override;
 
 private:
+  static constexpr char const *LOGGING_NAME = "PunishmentChannel";
```
Comment (n-hutton) [inline libs/muddle/include/muddle/punishment_broadcast_channel.hpp:118]: unrelated?

File: libs/semanticsearch/include/semanticsearch/schema/vocabulary_instance.hpp [hunk 17]
```diff
@@ -36,10 +36,7 @@ class VocabularyInstance : public std::enable_shared_from_this<VocabularyInstanc
   static Vocabulary New(T data)
   {
     // TODO(private issue 143): add destructor
-    Vocabulary ret;
-    ret.reset(new VocabularyInstance(std::type_index(typeid(T)), new T(data)));
-
-    return ret;
+    return Vocabulary{new VocabularyInstance(std::type_index(typeid(T)), new T(data))};  // NOLINT
```
Comment (n-hutton) [inline libs/semanticsearch/include/semanticsearch/schema/vocabulary_instance.hpp]: Has this snuck in and is causing build failures?

**Review by n-hutton — CHANGES_REQUESTED** [review]
Minor changes

File: libs/ledger/include/ledger/storage_unit/transaction_finder_protocol.hpp (hunk 4 above)
Comment (ejfitzgerald) [reply libs/ledger/include/ledger/storage_unit/transaction_finder_protocol.hpp:59]: No bug here, just the fact this was a generic protocol at one point and then became specialised but the interface was not updated. This change it mostly a code readability thing.

File: libs/ledger/include/ledger/storage_unit/transaction_store.hpp (hunk 8 above)
Comment (ejfitzgerald) [reply libs/ledger/include/ledger/storage_unit/transaction_store.hpp:55]: You mean the current subtree bug - nope I am planning on looking into that today. This code change is a restructuring of the existing code + that state machine bug fix (now with unit tests to catch it).

My plan is to look at the transaction syncing issue today

File: libs/ledger/include/ledger/storage_unit/transaction_store_aggregator.hpp (hunk 9 above)
Comment (ejfitzgerald) [reply libs/ledger/include/ledger/storage_unit/transaction_store_aggregator.hpp:33]: Done

File: libs/ledger/src/storage_unit/lane_service.cpp (hunk 10 above)
Comment (ejfitzgerald) [reply libs/ledger/src/storage_unit/lane_service.cpp]: Done

File: libs/semanticsearch/include/semanticsearch/schema/vocabulary_instance.hpp (hunk 17 above)
Comment (ejfitzgerald) [reply libs/semanticsearch/include/semanticsearch/schema/vocabulary_instance.hpp]: Yes it suddenly started failing the static analysis checks, not really sure why this is to be honest, but had to mark it as `NOLINT` because its suggestion (use `std::make_shared`) can't work because `VocabularyInstance` is private

File: libs/muddle/include/muddle/punishment_broadcast_channel.hpp (hunk 16 above)
Comment (ejfitzgerald) [reply libs/muddle/include/muddle/punishment_broadcast_channel.hpp:118]: Needed I removed the logging name from "Protocol" because this was being used as the default too often making the logging/debugging harder.

File: libs/ledger/tests/unit/storage/transaction_generator.hpp (hunk 15 above)
Comment (ejfitzgerald) [reply libs/ledger/tests/unit/storage/transaction_generator.hpp:51]: Not really - I did look into them but they seemed to be much more complicated than the above which is why I opted to have this simple subset for these tests

File: libs/ledger/tests/unit/storage/transaction_archiver_tests.cpp (hunk 14 above)
Comment (ejfitzgerald) [reply libs/ledger/tests/unit/storage/transaction_archiver_tests.cpp:61]: Sorry I can't remember, but if you look all `MATCHER` instances in the code base as it stands are `NOLINT`

File: libs/ledger/src/storage_unit/transaction_store_sync_protocol.cpp (hunk 13 above)
Comment (ejfitzgerald) [reply libs/ledger/src/storage_unit/transaction_store_sync_protocol.cpp:170]: There are NO major changes to TX sync. All the changes here are interface updates essentially. I will be looking into the main tx sync issue next

File: libs/ledger/include/ledger/storage_unit/recent_transaction_cache.hpp (hunk 1 above)
Comment (ejfitzgerald) [reply libs/ledger/include/ledger/storage_unit/recent_transaction_cache.hpp]: Done

File: libs/ledger/src/storage_unit/transaction_store_sync_protocol.cpp (hunk 13 above)
Comment (n-hutton) [reply libs/ledger/src/storage_unit/transaction_store_sync_protocol.cpp:170]: great

File: libs/muddle/include/muddle/punishment_broadcast_channel.hpp (hunk 16 above)
Comment (n-hutton) [reply libs/muddle/include/muddle/punishment_broadcast_channel.hpp:118]: Oh - that's an f up

**Review by n-hutton — APPROVED** [review]
Looks great!
